File: challenge_response/core/hand_action_detector.py
"""손 모양 판정 (SPEC 4.6).

패턴 매칭이 핵심이다. 손가락 4개(엄지 제외)의 펴짐/굽힘 조합이 정확히
일치할 때만 라벨을 준다. 단일 임계값으로 뭉뚱그리지 않기 때문에
NEG_threefingers([T,T,T,F])나 NEG_indexring([T,F,T,F])이 자연히 걸러진다.

임계값 리터럴은 없다. 전부 challenge_config.json에서 온다.
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from . import features as F
from . import geometry as g

UNKNOWN = "UNKNOWN"

# [index, middle, ring, pinky] 조합 → 라벨. 임계값이 아니라 프로토콜 정의다.
SHAPE_PATTERNS: dict[str, tuple[bool, bool, bool, bool]] = {
    "OPEN_PALM": (True, True, True, True),
    "FIST": (False, False, False, False),
    "INDEX": (True, False, False, False),
    "TWO_FINGERS": (True, True, False, False),
}


@dataclass(frozen=True)
class ShapeResult:
    label: str
    confidence: float
    flags: tuple[bool, ...]      # FINGER_NAMES 순서 (엄지 포함)
    angles: tuple[float, ...]    # FINGER_NAMES 순서
    margins: tuple[float, ...]   # 임계값과의 여유(도)


class HandActionDetector:
    """프레임 1개의 랜드마크 → (라벨, 신뢰도)."""

    def __init__(self, config: dict):
        angle_cfg = config["finger_extended_angle"]
        self.thumb_threshold = float(angle_cfg["thumb"])
        self.other_threshold = float(angle_cfg["others"])
        self.angle_space = str(config["angle_space"])
        self.confidence_margin_deg = float(config["shape_confidence_margin_deg"])

    def prepare(self, landmarks: np.ndarray, width: int | None = None,
                height: int | None = None) -> np.ndarray:
        """정규화 좌표를 설정된 angle_space로 옮긴다.

        angle_space가 "world"이면 MediaPipe world landmarks를 그대로 넘겨야 한다.
        """
        if self.angle_space == "world":
            return np.asarray(landmarks, dtype=np.float64)
        if width is None or height is None:
            raise ValueError("image_iso 좌표계는 width/height가 필요하다")
        return g.to_isotropic(landmarks, width, height)

    def detect(self, coords: np.ndarray) -> ShapeResult:
        """coords는 이미 self.angle_space 좌표계여야 한다 (prepare 참고)."""
        angles = g.finger_angle_array(coords)
        flags, margins = g.extension_flags(angles, self.thumb_threshold, self.other_threshold)

        idx = [g.FINGER_NAMES.index(n) for n in g.NON_THUMB_FINGERS]

        # 각도를 못 구한 손가락이 하나라도 있으면 판정하지 않는다.
        # 이게 없으면 랜드마크가 전부 nan인 프레임의 flags가 모두 False가 되어
        # FIST([F,F,F,F])로 오인된다.
        if not all(np.isfinite(angles[i]) for i in idx):
            return ShapeResult(UNKNOWN, 0.0, tuple(flags),
                               tuple(float(a) for a in angles), tuple(margins))

        pattern = tuple(flags[i] for i in idx)
        label = next((name for name, p in SHAPE_PATTERNS.items() if p == pattern), UNKNOWN)

        # 신뢰도: 판정에 쓰인 4손가락이 임계값에서 얼마나 확실히 떨어져 있는지.
        # 가장 아슬아슬한 손가락이 전체 신뢰도를 결정한다.
        scale = self.confidence_margin_deg
        decisiveness = [min(abs(margins[i]) / scale, 1.0) if np.isfinite(margins[i]) else 0.0
                        for i in idx]
        confidence = float(min(decisiveness)) if decisiveness else 0.0

        return ShapeResult(label=label, confidence=confidence,
                           flags=tuple(flags), angles=tuple(float(a) for a in angles),
                           margins=tuple(margins))

    def detect_from_clip(self, clip, frame_index: int) -> ShapeResult:
        """캐시된 영상의 한 프레임을 판정한다 (05_validate_rules.py용)."""
        coords = F.angle_source(clip, self.angle_space)[frame_index]
        return self.detect(coords)

File: lib/challenge/hand_action_detector.dart
/// 손 모양 판정. `challenge_response/core/hand_action_detector.py`의 이식.
///
/// 패턴 매칭이 핵심이다. 손가락 4개(엄지 제외)의 펴짐/굽힘 조합이 **정확히**
/// 일치할 때만 라벨을 준다. 단일 임계값으로 뭉뚱그리지 않기 때문에 세 손가락
/// [T,T,T,F]이나 검지+약지 [T,F,T,F] 같은 어중간한 손이 자연히 걸러진다.
///
/// 임계값 리터럴은 없다. 전부 [ChallengeConfig]에서 온다.
library;

import 'challenge_config.dart';
import 'geometry.dart';

const String kUnknownShape = 'UNKNOWN';

/// [index, middle, ring, pinky] 조합 → 라벨. 임계값이 아니라 프로토콜 정의다.
const Map<String, List<bool>> kShapePatterns = <String, List<bool>>{
  'OPEN_PALM': <bool>[true, true, true, true],
  'FIST': <bool>[false, false, false, false],
  'INDEX': <bool>[true, false, false, false],
  'TWO_FINGERS': <bool>[true, true, false, false],
};

class ShapeResult {
  final String label;

  /// 판정에 쓰인 4손가락이 임계값에서 얼마나 확실히 떨어져 있는지. 0.0~1.0.
  /// 가장 아슬아슬한 손가락이 전체 신뢰도를 결정한다.
  final double confidence;

  /// [kFingerNames] 순서(엄지 포함).
  final List<bool> flags;
  final List<double> angles;

  /// 임계값과의 여유(도).
  final List<double> margins;

  /// 손끝-손목 거리 / 손 크기. FIST 게이트에 쓴다.
  final double tipWristRatio;

  const ShapeResult({
    required this.label,
    required this.confidence,
    required this.flags,
    required this.angles,
    required this.margins,
    this.tipWristRatio = double.nan,
  });

  bool get isKnown => label != kUnknownShape;
}

class HandActionDetector {
  final double thumbThreshold;
  final double otherThreshold;
  final double confidenceMarginDeg;
  final String angleSpace;

  /// null이면 FIST 게이트를 걸지 않는다. OPEN_PALM 쪽에는 원래 걸지 않는다.
  final double? fistMaxTipWristRatio;

  HandActionDetector(ChallengeConfig config)
      : thumbThreshold = config.fingerExtendedAngle.thumb,
        otherThreshold = config.fingerExtendedAngle.others,
        confidenceMarginDeg = config.shapeConfidenceMarginDeg,
        angleSpace = config.angleSpace,
        fistMaxTipWristRatio = config.fistMaxTipWristRatio;

  /// 정규화 좌표를 설정된 [angleSpace]로 옮긴다.
  ///
  /// `world`는 MediaPipe world landmarks를 그대로 넘겨야 한다는 뜻인데,
  /// 앱이 쓰는 `hand_landmarker` 3.0.1은 world 좌표를 주지 않는다. 서버 설정이
  /// world로 바뀌면 조용히 틀린 각도로 판정하는 대신 여기서 막는다.
  Coords prepare(Coords landmarks, {int? width, int? height}) {
    if (angleSpace == 'world') {
      return landmarks;
    }
    if (width == null || height == null) {
      throw ArgumentError('image_iso 좌표계는 width/height가 필요하다');
    }
    return toIsotropic(landmarks, width, height);
  }

  /// [coords]는 이미 설정된 좌표계(image_iso면 종횡비 보정 완료)여야 한다.
  ShapeResult detect(Coords coords) {
    final List<double> angles = fingerAngles(coords);
    final r = extensionFlags(angles, thumbThreshold, otherThreshold);
    final List<int> idx = <int>[
      for (final String n in kNonThumbFingers) kFingerNames.indexOf(n),
    ];

    // 각도를 못 구한 손가락이 하나라도 있으면 판정하지 않는다.
    // 이게 없으면 랜드마크가 전부 NaN인 프레임의 flags가 모두 false가 되어
    // FIST([F,F,F,F])로 오인된다.
    final bool allFinite = idx.every((int i) => angles[i].isFinite);
    if (!allFinite) {
      return ShapeResult(
        label: kUnknownShape,
        confidence: 0.0,
        flags: r.flags,
        angles: angles,
        margins: r.margins,
      );
    }

    final List<bool> pattern = <bool>[for (final int i in idx) r.flags[i]];
    String label = kUnknownShape;
    for (final MapEntry<String, List<bool>> e in kShapePatterns.entries) {
      if (_samePattern(e.value, pattern)) {
        label = e.key;
        break;
      }
    }

    final double tipWrist = fingertipWristRatio(coords);
    if (label == 'FIST' && fistMaxTipWristRatio != null) {
      // 손끝 거리를 못 재면 반쯤 쥔 손인지 확인할 수 없으므로 FIST로 인정하지 않는다.
      if (!tipWrist.isFinite || tipWrist >= fistMaxTipWristRatio!) {
        label = kUnknownShape;
      }
    }

    double confidence = double.infinity;
    for (final int i in idx) {
      final double d = r.margins[i].isFinite
          ? (r.margins[i].abs() / confidenceMarginDeg).clamp(0.0, 1.0)
          : 0.0;
      if (d < confidence) confidence = d;
    }
    if (!confidence.isFinite) confidence = 0.0;

    return ShapeResult(
      label: label,
      confidence: confidence,
      flags: r.flags,
      angles: angles,
      margins: r.margins,
      tipWristRatio: tipWrist,
    );
  }

  static bool _samePattern(List<bool> a, List<bool> b) {
    for (int i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
}
